# APPROXIMATE SOLUTION OF THE SCHRÖDINGER EQUATION FOR A MODIFIED SQUARE WELL POTENTIAL - LIVE CODING SESSION (STUDENT HANDOUT)  #

# Background information and live coding exercise description

## Pre-requisite knowledge

**Scientific pre-requisite knowledge**

To complete the notebook below, students should be 

- familiar with the time-independent Schrödinger equation
- able to apply the postulates of quantum mechanics (operators, probabilistic interpretation of the wave function, orthogonality of functions, etc.) 
- familiar with the particle in a 1-D box (a.k.a. particle in an infinite well) model problem and the properties of the eigenfunctions
- able to use particle in a 1-D box eigenfunctions to calculate expectation values of operators (e.g. average kinetic energy) 

For an introduction to or review of the above topics, please see a standard Physical Chemistry text (e.g. sections 3.1 - 3.7, 4.1 - 4.3, and 4.5 in McQuarrie's [Physical Chemistry: A Molecular Approach](https://mitpress.mit.edu/9781940380216/physical-chemistry/) would serve as a good resource).

**Programming and Python pre-requisite knowledge**

To complete the notebook below, students should know

- basic Python syntax (variable types, variable assignment, formatted printing using f-strings, etc. )
- what functions are and how to use them
- how to load standard Python libraries (NumPy, SciPy, etc.) and use the functions included in these libraries
- how to write code for user-defined functions
- how to initialize matrices and vectors and how to access/modify their elements

For an introduction to or review of the above programming topics, please see the ["Introduction to programming and Python"](https://github.com/act-cms/lesson-01-introduction-to-programming-and-python.git) Jupyter notebook.

__A note on math background:__ Although familiarity with linear algebra concepts, such as matrix diagonalization, is helpful, they are not strictly necessary for you to complete this activity and develop an appreciation for the utility and key tenets of the linear variation method. [Section 0](#0.-eview-of-useful-properties-of-vectors-matrices-and-determinants) provides an overview of the basic properties of vectors, matrices, and determinants. Furthermore, if you are interested in answering questions like

- *why* calculating the kinetic and potential energy matrices is needed
- *how* diagonalizing the Hamiltonian leads to the set of parameters that minimizes the energy

I encourage you to read through the "simple" example given in the Appendix. The intent of the appendix is not to provide you with a rigorous exposition of the theory but rather to demonstrate how task of finding the lowest-energy solution leads to the matrix formulation of the time-independent Schrödinger equation and why matrix diagonalization is equivalent to solving for the optimal coefficients. Fear not if you do not have the energy or time to read through the appendix, you will still be able to follow the live coding activity and develop intuition for the linear variation method!

## Learning outcomes

After completing these exercises, you will be able to

- describe key aspects of the linear variation method
- apply the linear variational method to obtain approximate solutions of the time-independent Schrödinger equation
- explain the cost to benefit ratio of accuracy versus computational complexity
- demonstrate skills for debugging code designed for scientific applications
- access and modify elements of arrays and/or matrices
- use scipy library functions solve eigenvalue equations
- visualize the solutions of the time-independent Schrödinger equation

## <span style="color:blue"> Pre-lab activities </span>

<span style="color:blue">

Before coming to lab, please complete the following and **turn in your solutions to problems 2 - 7 at the beginning of the lab period**.

1. Read sections 0 through 4. *If you have some background in linear algebra (i.e. you know how to expand the determinant of a matrix or how to calculate a matrix vector product), you may skip section 0. If you want to see a detailed application of the linear variation method to a "simple" problem involving two basis functions, please see the appendix at the end of this notebook*
   
2. Provide a concise (3-4 sentence) summary of the basic tenets of the linear variation method.
   
3. Suppose you were to use the linear variation method with $N=4$ basis functions $\phi_n$ (where $\phi_n$ are particle in a box eigenfunctions) to calculate the ground-state energy for an electron that experiences the potential given in problem 4 above. Would you expect that your result for the ground-state energy is above or below the energy with two basis functions? Would the approximate ground-state energy be above or below the exact ground-state energy

4. Given that the particle in a box eigenfunctions $\phi_n(x) = \sqrt{\frac{2}{L}} {\rm sin} \left( \frac{n\pi x}{L} \right)$ are orthonormal eigenfunctions of the kinetic energy operator $\hat{T}_x = -\frac{\hbar^2}{2m} \frac{{\rm d}^2}{{\rm d}x^2}$, show that
\begin{equation}
T_{n,m} = \int_0^L \phi_n(x)^* \hat{T}_x \phi_m(x) {\rm d}x = 
\begin{cases}
0 \hspace{1.7cm} n \neq m\\
\\
n^2 \frac{\hbar^2\pi^2}{2mL^2} \hspace{0.55cm} n = m
\end{cases}
\end{equation}

5. A particle confined to a box of width $L$ experiences the potential
\begin{equation}
V(x) = \begin{cases}
 0 \hspace{2.1cm} 0 \leq x \leq L/2 \\
 \\
 V_0 \hspace{2cm} L/2 \leq x \leq L \\
 \\
 \infty \hspace{2cm} \text{everywhere else}
\end{cases} \tag{2.1}
\end{equation}
   Using the particle in a box eigenfunctions $\phi_n(x) = \sqrt{\frac{2}{L}} {\rm sin} \left( \frac{n\pi x}{L} \right)$, show that
\begin{equation}
V_{1,2} = \int_0^L \phi_1(x)^* \hat{V}\phi_{2}(x) {\rm d}x = -4.24 \hspace{1mm} {\rm eV}
\end{equation}
The integral
\begin{equation}
\int {\rm sin}(Ax){\rm sin}(Bx) dx =\frac{1}{2} \left( \frac{{\rm sin} [(A-B)x]}{A-B} - \frac{{\rm sin} [(A+B)x]}{A+B} \right) + C 
\end{equation}
may be useful here.
  
6. Show that the two vectors ${\bf c}^{(1)} = 
\begin{pmatrix}
 0.99860 \\ 0.03747
\end{pmatrix}$ and ${\bf c}^{(2)} = 
\begin{pmatrix}
- 0.03747 \\  0.99860
\end{pmatrix}$ are orthogonal and normalized. 

7. Consider $V(x)$ given at the beginning of [section 6](#6.-live-coding-exercise). Write down the integral expression(s) you will need to evaluate for computing the matrix element $V_{m,n}$ of the potential energy matrix $\bf V$. Include the appropriate limits of integration and other factors such as $V_0$. You do not need to solve the integrals, however, you should think about how you would apply your work from section 13 of the ["Introduction to programming and Python"](https://github.com/act-cms/lesson-01-introduction-to-programming-and-python) exercises to numerically calculate the integrals.
</span>

## 0. Review of useful properties of vectors, matrices, and determinants

This section provides a pragmatic review of properties of matrices and vectors. For a more in-depth discussion, please see a textbook on linear algebra.

<br> <center> **MATRICES** </center> <br>

An $N\times M$ matrix ${\bf A}$ is a 2-dimensional collection of $N\cdot M$ numbers
\begin{equation}
{\bf A} = 
\begin{pmatrix}
A_{1,1} & A_{1,2} & \dots  & A_{1,M} \\
A_{2,1} & A_{2,2} & \dots  & A_{2,M} \\
\vdots & \vdots   & \ddots & \vdots  \\
A_{N,1} & A_{N,2} & \dots  & A_{N,M}
\end{pmatrix}
\end{equation}
The transpose of a matrix (denoted by $^T$) is obtained by interchanging columns and rows. Thus
\begin{equation}
{\bf A}^T = 
\begin{pmatrix}
A_{1,1} & A_{2,1} & \dots  & A_{M,1} \\
A_{1,2} & A_{2,2} & \dots  & A_{M,2} \\
\vdots & \vdots   & \ddots & \vdots  \\
A_{1,N} & A_{2,N} & \dots  & A_{M,N}
\end{pmatrix}
\end{equation}
and the transpose of an $N\times M$ matrix is an $M\times N$ matrix. When $N=M$, the matrix is said to be a square matrix. Furthermore, if ${\bf A} = {\bf A}^{T}$ (i.e. $A_{n,m} = A_{m,n}$) the matrix ${\bf A}$ is said to be symmetric. A square matrix $\bf D$ is said to be diagonal when all off-diagonal elements ($D_{n,m}$ with $n\neq m$) are zero
\begin{equation}
{\bf D} = 
\begin{pmatrix}
D_{1,1} &     0    & \dots   & 0 \\
 0      & D_{2,2}  & \dots   & 0 \\
\vdots  & \vdots   & \ddots  & \vdots  \\
 0      & 0        & \dots   & D_{N,N}
\end{pmatrix}
\end{equation}

The identity matrix $\bf I$ is a special case of a diagonal matrix in which all diagonal elements are equal to 1
\begin{equation}
{\bf I} = 
\begin{pmatrix}
 1      &     0    & \dots   & 0 \\
 0      &     1    & \dots   & 0 \\
\vdots  & \vdots   & \ddots  & \vdots  \\
 0      & 0        & \dots   & 1
\end{pmatrix}
\end{equation}

A matrix $\bf A$ is said to invertible if there exists a matrix ${\bf A}^{-1}$ (called the inverse of ${\bf A}$) such that ${\bf A}^{-1} \bf A = I$. A matrix is invertible if its determinant (see below) does not vanish.

<br> <center> **VECTORS** </center> <br>

A column vector ${\bf c}$ is a 1-dimensional collection of $N$ numbers
\begin{equation}
{\bf c} = 
\begin{pmatrix}
c_1 \\
c_2 \\
\vdots \\
c_N
\end{pmatrix}
\end{equation}
where $c_i$ is the i$^{\rm th}$ element of the vector. Alternatively, we can view a colum vector as an $N \times 1$ matrix. The transpose of a column vector is a row vector (or a $1\times N$ matrix). The zero vector $\bf 0$ is an array of zeros. 

<br> <center> **DOT/SCALAR/INNER PRODUCTS OF VECTORS** </center> <br>

The dot product (a.k.a. inner product) of a row vector $\bf a$ with a colum vector $\bf b$ is a scalar
\begin{equation}
{\bf a \cdot b} = 
\begin{pmatrix}
a_1 & a_2 & \dots & a_N
\end{pmatrix}
\begin{pmatrix}
b_1 \\
b_2 \\
\vdots \\
b_N
\end{pmatrix}=
a_1b_1 + a_2b_2 + \dots +a_Nb_N = \sum_{n}^N a_nb_n
\end{equation}
Note that for the dot product to be well-defined, the length of the two vectors must be the same. As an example, the dot product of
\begin{equation}
{\bf a} = 
\begin{pmatrix}
0 & 1 & 2  \\
\end{pmatrix}
\hspace{2cm}
{\bf b} = 
\begin{pmatrix}
3 \\
4 \\
5
\end{pmatrix}
\end{equation}
is
\begin{equation}
{\bf c} = 
\begin{pmatrix}
0 & 1 & 2  \\
\end{pmatrix}
\begin{pmatrix}
3 \\
4 \\
5
\end{pmatrix} =
0 \cdot 3 + 1 \cdot 4 + 3 \cdot 5 = 19
\end{equation}
The norm $\vert\vert {\bf c} \vert \vert $ of a vector $\bf c$ is $\vert\vert {\bf c} \vert \vert = \sqrt{{\bf c \cdot c}}$.

<br> <center> **MATRIX VECTOR PRODUCTS** </center> <br>

The product of an $N\times M$ matrix $\bf A$ with a $M\times 1$ column vector $\bf b$ is an $N\times 1$ column vector $\bf c$ with elements
\begin{equation}
c_n = \sum_m^M A_{n,m} b_m
\end{equation}
\begin{equation}
{\bf Ab} = 
\begin{pmatrix}
A_{1,1} & A_{1,2} & \dots  & A_{1,M} \\
A_{2,1} & A_{2,2} & \dots  & A_{2,M} \\
\vdots & \vdots   & \ddots & \vdots  \\
\vdots & \vdots   & \ddots & \vdots  \\
A_{N,1} & A_{N,2} & \dots  & A_{N,M}
\end{pmatrix}
\begin{pmatrix}
b_1 \\
b_2 \\
\vdots \\
b_M
\end{pmatrix} =
\begin{pmatrix}
A_{1,1}b_1 + A_{1,2}b_2 + \dots + A_{1,M}b_M \\
A_{2,1}b_1 + A_{2,2}b_2 + \dots + A_{2,M}b_M \\
\vdots \\
\vdots \\
A_{N,1}b_1 + A_{N,2}b_2 + \dots + A_{N,M}b_M
\end{pmatrix} =
\bf{c}
\end{equation}
Note that for the matrix-vector to be well-defined, the length of the vector and the number of columns in the matrix must be the same. Looking at it a different way, the $n^{\rm th}$ element of the vector $\bf c$ is the dot product between the $n^{\rm th}$ row of $\bf A$ and the vector $\bf b$.
As an example, matrix vector product of
\begin{equation}
{\bf A} = 
\begin{pmatrix}
0 & 1  \\
2 & 3  \\
4 & 5
\end{pmatrix}
\hspace{2cm}
{\bf b} = 
\begin{pmatrix}
6  \\
7 
\end{pmatrix}
\end{equation}
is
\begin{equation}
{\bf c} = 
\begin{pmatrix}
0 & 1  \\
2 & 3  \\
4 & 5
\end{pmatrix}
\begin{pmatrix}
6  \\
7 
\end{pmatrix} =
\begin{pmatrix}
0\cdot 6 + 1\cdot 7 \\
2\cdot 6 + 3\cdot 7 \\
4\cdot 6 + 5\cdot 7 \\
\end{pmatrix}=
\begin{pmatrix}
7  \\
33 \\
59
\end{pmatrix}
\end{equation}

<br> <center> **MATRIX MATRIX PRODUCTS** </center> <br>

The matrix product of an $N\times K$ matrix $\bf A$ and a $K\times M$ matrix $\bf B$ is an $N\times M$ matrix $\bf C$ with elements
\begin{equation}
C_{n,m} = \left( {\bf AB} \right)_{n,m} = \sum_{k=1}^{K} A_{n,k}\cdot B_{k,m}
\end{equation}
Note that for the matrix product to be well-defined, the number of columns in matrix $\bf A$ and the number of rows in matrix $\bf B$ must be the same. 
As an example, the matrix product of 
\begin{equation}
{\bf A} = 
\begin{pmatrix}
0 & 1 \\
2 & 3
\end{pmatrix}
\hspace{2cm}
{\bf B} = 
\begin{pmatrix}
4 & 5 & 6 \\
7 & 8 & 9
\end{pmatrix}
\end{equation}
is given by
\begin{equation}
{\bf C = AB} =
\begin{pmatrix}
0 & 1 \\
2 & 3
\end{pmatrix}
\begin{pmatrix}
4 & 5 & 6 \\
7 & 8 & 9
\end{pmatrix} =
\begin{pmatrix}
0\cdot 4 + 1\cdot 7 & 0\cdot 5 + 1\cdot 8 & 0\cdot 6 + 1\cdot 9 \\
2\cdot 4 + 3\cdot 7 & 2\cdot 5 + 3\cdot 8 & 2\cdot 6 + 3\cdot 9 \\
\end{pmatrix}=
\begin{pmatrix}
7  &  8 &  9 \\
29 & 34 &  39
\end{pmatrix}
\end{equation}
Matrix products are associative ($\bf A(BC) = (AB)C$), distributive ($\bf A(B+C) = AB + AC$), and in general (though not always) non-commutative ($\bf AB \neq BA$). Note that the identity matrix commutes with any other matrix $\bf IA = AI$ and the product of the identiy matrix with another matrix is the matrix itself ($\bf IA = A$). 

<br> <center> **DETERMINANTS** </center> <br>

The detemrminant of an $N \times N$ square matrix $\bf A$, denoted ${\rm det} ({\bf A})$ or $\vert {\bf A} \vert$, is a scalar and is the weighted sum of the determinants of the minors of $\bf A$. The minor ${\bf m}_{i,j}$ of $\bf A$ is an $N-1 \times N-1$ matrix obtained by removing the $i^{\rm th}$ row and $j^{\rm th}$ column from $\rm A$. Specifically, for every row $i$ of $\bf A$ the Laplace expansion is
\begin{equation}
\vert {\bf A} \vert = \sum_{j=1}^N \left( -1 \right)^{i+j}A_{i,j}\vert {\bf d}_{i,j} \vert .
\end{equation}
As an example, the determinant of
\begin{equation}
{\bf A} =
\begin{pmatrix}
1 & 2 & 3 \\
4 & 5 & 6 \\
7 & 8 & 9
\end{pmatrix}
\end{equation}
is obtained by choosing the first row ($i=1$) and expanding as
\begin{equation}
\vert {\bf A} \vert =  
\begin{vmatrix}
1 & 2 & 3 \\
4 & 5 & 6 \\
7 & 8 & 9
\end{vmatrix} = 
\left( -1 \right)^{1+1} \cdot 1 \cdot \begin{vmatrix} 5 & 6 \\ 8 & 9 \end{vmatrix} + \left( -1 \right)^{1+2} \cdot 2 \cdot \begin{vmatrix} 4 & 6 \\ 7 & 9 \end{vmatrix} + \left( -1 \right)^{1+3} \cdot 3 \cdot \begin{vmatrix} 4 & 5 \\ 7 & 8 \end{vmatrix} = 1 \cdot \begin{vmatrix} 5 & 6 \\ 8 & 9 \end{vmatrix} - 2 \cdot \begin{vmatrix} 4 & 6 \\ 7 & 9 \end{vmatrix} + 3 \cdot \begin{vmatrix} 4 & 5 \\ 7 & 8 \end{vmatrix}  
\end{equation}
Next, we expand the remaining determinants to get the final result
\begin{equation}
\vert {\bf A} \vert = 
1 \cdot \begin{vmatrix} 5 & 6 \\ 8 & 9 \end{vmatrix} - 2 \cdot \begin{vmatrix} 4 & 6 \\ 7 & 9 \end{vmatrix} + 3 \cdot \begin{vmatrix} 4 & 5 \\ 7 & 8 \end{vmatrix} = 
1 \cdot \left( 5 \cdot 9 - 6 \cdot 8 \right) -2 \cdot \left( 4 \cdot 9 - 6 \cdot 7 \right) + 3 \cdot \left( 4 \cdot 8 - 5 \cdot 7 \right) = 0
\end{equation}

Note that we get the same result if we expand along any of the rows or any of the columns of $\bf A$. Furthermore, the sign of the determinant changes if one interchanges any two rows or any two columns of $\bf A$. The determinant of a product of two matrices $\bf A$ and $\bf B$ is the product of the determinants of the matrices; $\vert {\bf AB}\vert = \vert {\bf A}\vert \cdot \vert {\bf B}\vert$. Lastly, the determinant of a diagonalizable matrix $\bf A$ is equal to the product of the eigenvalues $\{ e_n \}$; $\vert {\bf A} \vert = \Pi_n e_n$.



## 1. Overview of the linear variation method

Analytical solutions to the time-independent Schrödinger equation
\begin{equation}
\hat{H} \psi = E \psi
\tag{1.1}
\end{equation}
are known for a select few model problems (e.g. particle in a 1-D/2-D/3-D box, harmonic oscillator, particle on a ring/sphere, hydrogen atom and one-electron ions), and most problems encountered in quantum chemistry necessitate that we make some sort of approximation when solving the time-independent Schrödinger equation. 

One such approximation method, called the *linear variational method*, expresses the wave function $\psi$ as a *linear combination* of *basis functions* $\{ \phi_n \}$
\begin{equation}
\psi = \sum_{n=1}^{N} c_n \phi_n
\tag{1.2}
\end{equation}
with *constant coefficients* $\{ c_n \}$. Put "simply", the basis functions are used to describe the wave function, and the importance of a basis function $\phi_n$ in this description is proportional to the magnitude of the coefficient $c_n$. For pratical reasons, the number of basis functions $N$ is finite; thus, the wave function in Eq. (1.2) is, in general, approximate. Because the wave function depends on the coefficients, the energy also depends on the coefficients. Thus, to find the ground-state solution to the time-independent Schrödinger equation, our task is to determine the set of coefficients that minimize the energy. Once we find the optimal set of normalized coefficients, we can express the wave function as in Eq. (1.2) and calculate the approximate energy as $E_{approx}=\int \psi^* \hat{H} \psi {\rm d}\tau$ (where ${\rm d} \tau$ denotes a general volume element) and other expectation values of interest. For the ground state (and in some cases excited states) one can prove that the approximate energy is guaranteed to be an *upper bound* to the exact energy (i.e. $E_{approx} \geq E_{exact}$ the approximate energy approaches the exact energy from above as the number of basis functions is increased).

 [Section 2](#general-approach-to-using-the-linear-variation-method) presents the general approach to applying the linear variation method. [Section 3](#a-note-on-atomic-units) introduces atomic units, and the basic syntax for matrix diagonalization is presented in [section 4](#matrix-diagonalization-in-python). In the [appendix](#appendix:-a-simple-example-to-demonstrate-the-application-of-the-linear-variation-method), we use a "simple" problem to demonstrate the application of the linear variation method. During this discussion, we highlight concepts such as the matrix representation of the Hamiltonian, matrix form of the time-independent Schrödinger equation, eigenvalues, eigenvectors, and matrix diagonalization.

## 2. General approach to applying the linear variation method

To obtain approximate solutions to the time-indpendent Schrödinger equation within the framework of the linear variation method, one can follow the following steps.

1. Select a basis set $\{\phi_n\}$ and determine the number of basis functions ($N$) to include in the calculation.
   - The basis functions should obey the appropriate boundary conditions
   - Although the basis functions do not necessarily have to be orthonormal, it helps when they are.
\begin{equation}
\int_\text{all space} \phi_n^*  \phi_m {\rm d}\tau = \delta_{m,n}
\end{equation}
where the Krönecker delta function $\delta_{m,n}$ is defined as 

\begin{equation}
\delta_{n,m}= \begin{cases}
 0 \hspace{2.0cm} n \neq  m\\
 \\
 1 \hspace{2cm} n = m\\
\end{cases}. 
\end{equation}

<br>

2. Calculate the matrix representation of Hamiltonian (an $N\times N$ matrix)
\begin{equation}
\bf H = T + V \tag{3.1}
\end{equation}
In the case of an orthonormal basis, the matrix form of the time-independent Schrödinger equation is ${\bf Hc} = E {\bf c}$.
   - Since the Hamiltonian is the sum of the kinetic ($\hat{T}$) and potential ($\hat{V}$) energy operators, computing the matrix representation the kinetic ($\bf T$) and potential ($\bf V$) energy matrices separately helps organize and debug your code.
   - The kinetic energy matrix elements are given by
\begin{equation}
T_{n,m} = \int_\text{ all space} \phi_n^* \hat{T} \phi_m {\rm d}\tau
\tag{3.2}
\end{equation}
   - The potential energy matrix elements are given by
\begin{equation}
V_{n,m} = \int_\text{all space} \phi_n^* \hat{V} \phi_m {\rm d}\tau
\tag{3.3}
\end{equation}
   - Depending on the operator and the basis set, these integrals may be simplified. If they cannot, one can (hopefully) use some numerical integration and/or differentiation technique.
     
<br>

3. Diagonalize the Hamiltonian to obtain the eigenvalues (approximate energies) and eigenvectors (expansion coefficients). There will be $N$ eigenvalues $\{ E^{(n)}\}$ and $N$ eigenvectors $\{ {\bf c}^{(n)}\}$ (column vectors). Generally, the eigenvectors after diagonalization are orthonormal

\begin{equation}
\left({\bf c}^{(n)}\right)^T \cdot {\bf c}^{(m)} = \delta_{m,n} 
\end{equation}

4. Using the expansion coefficients, the approximate wave function $\psi^{(n)}$ corresponding to eigenvalue $E^{(n)}$ 
\begin{equation}
\psi^{(n)} = \sum_{i=1}^{N} c_i^{(n)} \phi_i
\end{equation}
can be used to calculate properties of the system.

## 3. A note on atomic units

Because expressing properties (e.g. energies and distances) of atomic systems is quite cumbersome in SI units, most calculations on the atomic scale use *atomic units*. In atomic units,
\begin{equation}
m_e = e = 4\pi\varepsilon_0 = \hbar = a_0 = E_h = 1,
\end{equation}
where $1 \hspace{1mm} a_0 = \frac{4\pi\varepsilon_0 \hbar^2}{m_ee^2}=0.529177 $ ${\mathring{\rm{A}}}$ is the Bohr radius and $1 \hspace{1mm} E_h = \frac{\hbar^2}{m_ea_0^2} = 27.2114$ eV is the Hartree. Note that although the mass of an electron is 1 in atomic units, the mass of a proton is much larger
\begin{equation}
m_p^{(au)} = \frac{m_p^{(kg)}}{m_e^{(kg)}} \cdot m_e^{(au)} = \frac{1.67262\times10^{-27}}{9.10939\times 10^{-31}} = 1836.15
\end{equation}

## 4. Matrix diagonalization in Python

As noted in sections 2 and 3, determining the ground-state energy and corresponding wave function requires the diagonalization of the Hamiltonian matrix. Python's scipy libraries offer a variety of functions to solve eigenvalue problems. Because the Hamiltonian matrix is symmetric and we want both the eigenvectors and eigenvalues, we will use the ```eigh``` function. The basic syntax for diagonalizing an $N\times N$ matrix $\bf M$ is

```Python
eig_vals , eig_vecs = scipy.linalg.eigh(M)
```
where, upon return, the vector ```eig_vals``` contains the $N$ eigenvalues of the matrix $\bf M$ in **ascending order**, and the columns of the $N\times N$ matrix ```eig_vecs``` correspond to the eigenvectors. That is, the eigenvector associated with the $i^{\rm th}$ eigenvalue (```eig_vals[i]```) of $\bf M$ is in the $i^{\rm th}$ column of the matrix ```eig_vecs``` (```eig_vecs[ :, i ] ```). To find out more about the advanced options for ```eigh```, use ```help(scipy.linalg.eigh)```.

To demonstrate the syntax, let's diagonalize the Hamiltonian matrix given in Eq. A.16 of the Appendix.

In [1]:
import numpy as np
import scipy.linalg as spla

# define Hamiltonian matrix based on the appendix
H = np.array( [ [ 42.6, -4.24 ] , [ -4.24 , 155.4] ] )

# initialize arrays/matrices to store eigenvalues and eigenvectors (the size of the Hamiltonian is 2 x 2, so N = 2)
N = 2
eig_vals = np.array( N )
eig_vecs = np.array( ( N , N ) )

# diagonalize the Hamiltonian matrix
eig_vals , eig_vecs = spla.eigh( H )

# define an array for printing results (strictly speaking, this is not necessary, 
# but it helps in demonstrating how to access eigenvectors
vec_print = np.zeros( N )

# print the eigenvalues and corresponding eigenvectors

# loop over eigenpairs
for i in range( N ):

    e_print   = eig_vals[i]
    vec_print = eig_vecs[ :, i ]
    
    print(F'State index={i:2d}   Energy(eV)={e_print:8.3f}  Eigenvector=( {vec_print[0]:8.4f} , {vec_print[1]:8.4f} )')

# demonstrate the eigenvectors are orthonormal

S_matrix = np.array( ( N , N ) )
S_matrix = np.matmul( eig_vecs, np.transpose( eig_vecs ) )

print()
print("The overlap matrix between the eigenvectors is (within machine precision, this should be the identity matrix):")
print()
print(S_matrix)

State index= 0   Energy(eV)=  42.441  Eigenvector=(  -0.9993 ,  -0.0375 )
State index= 1   Energy(eV)= 155.559  Eigenvector=(  -0.0375 ,   0.9993 )

The overlap matrix between the eigenvectors is (within machine precision, this should be the identity matrix):

[[1.00000000e+00 1.97694227e-18]
 [1.97694227e-18 1.00000000e+00]]


## 5. Python libraries for physical constants

To avoid typographical errors, it is good practice use standardized libraries for retrieving physical constants. SciPy's ```constants``` library provides a variety of physical constants such as $h$, $c$, $N_A$, etc. To load the library, you can use the syntax

```python
import scipy.consts as spconst
```

To use a physical constant denoted with ```symbol_for_physical_constant```, you can use the syntax ```spconst.symbol_for_physical_constant```; for example, to use the speed of light in vacuum, you would use ```sponst.c```. Of course, the units of $c$ are important. Although it is a fairly safe assumption that SciPy most likely uses SI units (m$\cdot$s$^{-1}$ for $c$), it is good to check. To do so, we can use the following syntax

```python
print(spconst.physical_constants['description_of_physical_constant'][1])
```

Technically, ```spconst.physical_constants['description_of_physical_constant']``` returns an array of length 3. The 0$^{\rm th}$ element represent the value of the physical constant and the first element indicates the units in which the constant is expressed. Going back the the speed of ight in vacuum example, the command ```print(spconst.physical_constants['speed of light in vacuum'][1])``` would indicate that the value of $c$ is indeed expressed in units m$\cdot$s$^{-1}$.  

To find out what physical constants are available and what the symbols and descriptions are for the physical constants, you can use```help(spconst)```. The same command will also give you more information on SciPy's physical constants library if you are interested.

## <span style="color:green"> 6. Live coding activity </span>

<span style="color:green">

In this exercise, we will consider a particle with mass $M$ that moves along $x$ and experiences the potential
\begin{equation}
V(x) = \begin{cases}
 0 \hspace{2.1cm} 0 \leq x \leq L/2 \\
 \\
 V_0 \hspace{2cm} L/2 \leq x \leq L \\
 \\
 \infty \hspace{2cm} \text{everywhere else}
\end{cases}
\end{equation}
where $V_1 > 0$. For a visualization of the potential, see the image below.

<img src="https://github.com/act-cms/lesson-02-linear-variation-method/blob/main/potential_live_coding.jpg?raw=true" style="display: block; margin: 0 auto; max-height:200px;">

Using the orthonormal particle in 1-D box eigenfunctions
\begin{equation}
\phi_n(x) = \sqrt{\frac{2}{L}} {\rm sin}\left( \frac{n\pi x}{L}\right) \hspace{1cm} n = 1,2, ...
\end{equation}
as our basis functions, we will demonstrate how to implement the linear variational method. To do so, we will write code that

1. builds the kinetic energy matrix (see Eq. 3.2 and pre-lab exercise 4)
2. builds the potential energy matrix (see Eq. 3.3 and pre-lab exercise 7)
3. builds the Hamiltonian from the kinetic and potential energy matrices (see equation 3.1)
4. calculates the approximate solutions (energies and expansion coefficients) of the time-independent Schrödinger equation by diagonalizing the Hamiltonian (see [section 4](#matrix-diagonalization-in-python))
5. calculates the value of the ground-state wave function at some value of $x$ (see section 12 of the ["Introduction to programming and Python"](https://github.com/act-cms/lesson-01-introduction-to-programming-and-python) exercises)
6. Prints important information and key results including the input parameters ($N$, $W$s, $V$s, etc ...) and the ground-state energy.

In our algorithm, we will

- keep $M$, $L$, $V_1$, and $N$ (the number of basis functions in the linear expansion in Eq. (1.2)) as input variables that the user can specify
- assume that the user will input distances in units of  ${\mathring{\rm{A}}}$, energies in units of eV, and masses in units of kg
- use atomic units throughout the calculations
- to simplify debugging and to increase readability of your code, use functions as appropriate

Once we have implemented our algorithm, we will demonstrate approaches to check our results are sensible and that our code gives the correct results for limiting cases (e.g. does our algorithm give the right result in the limit when $V_1 = 0$ or when $V_1$ is very large?). 

# Python code development

## Load libraries

## Define input variables 

## Define functions

In this section, we will add as many dropdown cells as needed to implement the functions that are going to be used in the code. 

__NOTE:__ To aid in reading and debugging your code, use a separate cell for each function! Lastly, to aid the end-user, make sure to include comments and documentation in the form of DocStrings for your functions. For a refresher on how to use DocStrings, see section 8 of the ["Introduction to programming and Python"](https://github.com/act-cms/lesson-01-introduction-to-programming-and-python) exercises.

In [78]:
# if needed, write code for a function here

In [79]:
# if needed, write code for a function here

In [80]:
# if needed, write code for a function here

In [81]:
# if needed, write code for a function here

In [82]:
# if needed, write code for a function here

In [94]:
# if needed, write code for a function here

In [84]:
# if needed, write code for a function here

In [85]:
# if needed, write code for a function here

In [86]:
# if needed, write code for a function here

## Initialize arrays and matrices 

## Convert parameter values to atomic units

## Calculate the kinetic energy, potential energy, and finally, the Hamiltonian matrices

## DIagonalize the Hamiltonian to calculate the approximate solutions to the Schrödinger equation

## Print key results

# Appendix

## A simple example to demonstrate the application of the linear variation method

To simplify the notation, we will **assume that the coefficients, basis functions, and hence the wave function are real** for the remainder of this section. In most numerical applications, this turns out to be the case. Suppose we are interested in solving the time-independent Schrödinger equation (Eq. (1.1)) for an electron that experiences the potential
\begin{equation}
V(x) = \begin{cases}
 0 \hspace{2.1cm} 0 \leq x \leq L/2 \\
 \\
 V_0 \hspace{2cm} L/2 \leq x \leq L \\
 \\
 \infty \hspace{2cm} \text{everywhere else}
\end{cases} \tag{A.1}
\end{equation}

For a visualization of the potential, see the image below.

<img src="https://github.com/act-cms/lesson-02-linear-variation-method/blob/main/potential_live_coding.jpg?raw=true" style="display: block; margin: 0 auto; max-height:200px;">

This potential effectively confines the particle to $0 \leq x \leq L$ and the electron experiences a higher potential in the right half of the box. Although the potential looks deceptively simple, the analytical solutions to Eq. (1.1) are not known. Below, we present a detailed discussion of obtaining the approximate solutions using the linear variation method.

To start, we select a set of basis functions; because the potential in Eq. (A.1) is similar to the particle in a 1-D particle in a box potential, we will choose the eigenfunctions of the particle in a 1-D box Hamiltonian
\begin{equation}
\phi_n(x) = \sqrt\frac{2}{L}{\rm sin}\left( \frac{n\pi x}{L}\right)
\tag{A.2}
\end{equation}
as our basis functions. Although other choices are also possible (as long as they obey the right boundary conditions), the basis functions in Eq. (A.2) are convenient since they are orthonnormal (you showed this in section 13 of the ["Introduction to programming and Python"](https://github.com/act-cms/lesson-01-introduction-to-programming-and-python) exercises) on the domain $0 \leq x \leq L$. That is,

\begin{equation}
\int_0^L \phi_n(x)^*  \phi_m(x) {\rm d}x = \delta_{m,n}
\tag{A.3}
\end{equation}

where the Krönecker delta function $\delta_{m,n}$ is defined as 

\begin{equation}
\delta_{n,m}= \begin{cases}
 0 \hspace{2.0cm} n \neq  m\\
 \\
 1 \hspace{2cm} n = m\\
\end{cases}. 
\end{equation}

Having chosen the basis set, we need to decide how many basis functions to consider when describing the wave function. For simplicity, we will use two functions ($N=2$), and thus our wave function has the form
\begin{equation}
\psi(x) = c_1 \phi_1(x) + c_2 \phi_2(x) .
\tag{A.4}
\end{equation}

To determine the equations for the coefficients that minimize the energy, we first multiply Eq. 1.1 from the left by $\psi(x)^*$ and integrate
\begin{equation}
\int_{0}^{L} \psi(x)^* \hat{H} \psi(x) dx = E \int_{0}^L \psi(x)^* \psi(x) dx
\tag{A.5}
\end{equation}
Note that we did not assume that $\psi$ is normalized and kept the integral on the right hand side. Inserting Eq. (A.4) into Eq. (A.5), recognizing that the basis functions are orthonormal (as in Eq. (A.3)), and assuming that the coefficients are real, we have
\begin{equation}
H_{1,1}c_1^2 + H_{1,2} c_1c_2 + H_{2,1}c_1c_2 + H_{2,2}c_2^2 = E \left ( c_1^2 + c_2^2\right),
\tag{A.6}
\end{equation}
where the Hamiltonian matrix element $H_{n,m}$ is defined as
\begin{equation}
H_{n,m} = \int_{0}^L \phi_n(x) \hat{H} \phi_m(x){\rm d}x .
\tag{A.7}
\end{equation}
Note that $H_{n,m}$ is independent of the coefficients and can be evaluated once the form of $\hat{H}$ and $\phi_m$ are known. 
Finally, since the energy is stationary with respect to variations in the coefficients $\left( \text{i.e.} \frac{\partial E}{\partial c_1} = \frac{\partial  E}{\partial c_2} = 0\right) $ at the minimum, differentiating Eq. (A.6) with respect to $c_1$ and $c_2$ and recognizing that $H_{1,2} = H_{2,1}$ gives the two equations
\begin{equation}
H_{1,1} c_1 + H_{1,2}c_2 = Ec_1 
\tag{A.8a}
\end{equation}
\begin{equation}
H_{2,1} c_1 + H_{2,2}c_2 = Ec_2 
\tag{A.8b}
\end{equation}
In principle, one could solve this system of two equations by substitution, to get a quadratic equation for the energy $E$. Once the two roots are found, the coefficients corresponding to each root can be found from Eqs. (A.8a) and (A.8b).

However, this process becomes impractical very quickly for $N>2$ and we take a different approach. Notice that we can succinctly represent Eqs. (A.8a) and (A.8b) as one *matrix equation*

\begin{equation}
\begin{pmatrix}
H_{1,1} & H_{1,2} \\
H_{2,1} & H_{2,2} \\
\end{pmatrix}
\begin{pmatrix}
c_1 \\
c_2
\end{pmatrix}
= E
\begin{pmatrix}
c_1 \\
c_2
\end{pmatrix}
\iff {\bf H}{\bf c} = E{\bf c}
\tag{A.9} 
\end{equation}
Equation (A.9) is referred to as the *matrix form* of the time-independent Schrödinger equation in an orthonormal basis and $\bf H$ is called the matrix representation of the Hamiltonian (or Hamiltonian matrix for short) in the chosen basis set. The vector ${\bf c}$ contains the expansion coefficients and is referred to as the eigenvector of ${\bf H}$. $E$ is the eigenvalue that corresponds to the eigenvector ${\bf c}$. Collectively, ${\bf c}$ and $E$ are called an eigenpair.

Now, to solve for the coefficients, we multiply the right side of Eq. (A.9) by the identity matrix ${\bf I} = \begin{pmatrix} 1 & 0 \\ 0 & 1\end{pmatrix}$ and subtract the result from both sides to give
\begin{equation}
\left( {\bf H} - E{\bf I} \right) {\bf c} = {\bf 0}
\tag{A.10}
\end{equation}
The trivial solution ${\bf c} = {\bf 0} = \begin{pmatrix} 0 \\ 0 \end{pmatrix}$ is not interesting since it implies that $\psi = 0$. The more interesting solution is obtained by considering properties of ${\bf H} - E{\bf I}$. If it is invertible $\left( \text{i.e.} \left( {\bf H} - E{\bf I}\right)^{-1} \left({\bf H} - E{\bf I}\right) = {\bf I} \right)$, multiplying Eq. (A.10) from the left by $\left({\bf H} - E{\bf I}\right)^{-1}$ implies that $\bf c = 0$
\begin{equation}
\left( {\bf H} - E{\bf I}\right)^{-1} \left({\bf H} - E{\bf I}\right) {\bf c} = \left( {\bf H} - E{\bf I}\right)^{-1} {\bf 0} \hspace{1cm}\Rightarrow\hspace{1cm} {\bf I}{\bf c} = {\bf 0} \hspace{1cm}\Rightarrow\hspace{1cm}  {\bf c} = {\bf 0}
\end{equation}
Thus, the inverse of ${\bf H} - E{\bf I}$ must not exist for non-trivial solutions (i.e. ${\bf c} \neq {\bf 0}$). This condition is met, if the determinant of ${\bf H} - E{\bf I}$ vanishes
\begin{equation}
\vert {\bf H} - E{\bf I} \vert = 0
\tag{A.11}
\end{equation}

Expanding the determinant and solving for the roots of the resulting polynomial gives the eigenvalues of ${\bf H}$. Then, for each eigenvalue, we can determine the associated eigvector ${\bf}$ from Eq. (A.9). 

To demonstrate this process, we first calculate the Hamiltonian matrix elements using Eqs (A.1), (A.2), and (A.7). Since the Hamiltonian operator is the sum of the kinetic energy operator ($\hat{T}_x$) and the potential energy opertor, the matrix representation of the Hamiltonian is $\bf H = T + V$. Since the basis functions in Eq. (A.2) are orthonormal eigenfunctions of the particle in a 1-D box Hamiltonian (for which $\hat{H} = \hat{T}_x = - \frac{\hbar^2}{2m}\frac{d^2}{dx^2}$) operator, the elements of the kinetic energy matrix are 
\begin{equation}
T_{n,m} = m^2 \left(\frac{\hbar^2\pi^2}{2m_eL^2}\right) \delta_{n,m}.
\tag{A.12}
\end{equation}

Since the value of the Krönecker delta function $\delta_{n,m}$ vanishes unless $n=m$ and is equal to 1 when $n=m$, we see that $\bf T$ is a diagonal matrix. For an electron with mass $m_e$ in a box of width $L=1$ ${\mathring{\rm{A}}}$, $ \left(\frac{\hbar^2\pi^2}{2m_eL^2}\right) = 37.6$ eV, and thus, $\bf T$ (in units of eV) has the form
\begin{equation}
{ \bf T} =
\begin{pmatrix}
37.6 & 0 \\
0  & 150.4
\end{pmatrix}
\tag{A.13}
\end{equation}
Now, since $V(x)=0$ for $0\leq x \leq L/2$ and $V = V_0$ for $L/2 \leq x \leq L$, the potential energy matrix element
\begin{equation}
V_{n,m} = \int_0^L \phi_n(x)\hat{V}(x)\phi_m(x) {\rm d}x = \int_{L/2}^L \phi_n(x)V_0\phi_m(x) {\rm d}x = V_0\int_{L/2}^L \phi_n(x)\phi_m(x) {\rm d}x
\tag{A.14}
\end{equation}
requires that we evaluate the integral in Eq. (A.14). For this case, this can be done by hand, however, one needs to use numerical integration techniques for the general case. For  $V_0 = 10$ eV, the potential energy (in units of eV) is
\begin{equation}
{ \bf V} =
\begin{pmatrix}
5.0 & -4.24 \\
-4.24  & 5.0
\end{pmatrix}
\tag{A.15}
\end{equation}
Thus, the Hamiltonian matrix is
\begin{equation}
{ \bf H} =
\begin{pmatrix}
42.6 & -4.24 \\
-4.24 & 155.4
\end{pmatrix}
\tag{A.16}
\end{equation}
and hence ${\bf H} - E {\bf I}$ is given by
\begin{equation}
{ \bf H} - E{\bf I} =
\begin{pmatrix}
42.6 & -4.24 \\
-4.24 & 155.4
\end{pmatrix}
- E
\begin{pmatrix}
1 & 0 \\
0 & 1
\end{pmatrix}
= \begin{pmatrix}
42.6 - E & -4.24 \\
-4.24 & 155.4 - E 
\end{pmatrix}
\tag{A.17}
\end{equation}
To find the eigenvalues $E$, we expand the determinant
\begin{equation}
\vert {\bf H} - E{\bf I} \vert = 0
\hspace{0.5cm}\Rightarrow\hspace{0.5cm}
\begin{vmatrix}
42.6 -E & -4.24 \\
-4.24 & 155.4 - E
\end{vmatrix} =0
\hspace{0.5cm}\Rightarrow\hspace{0.5cm}
(42.6 - E)(155.4 - E) - (-4.24)^2 = 0
\hspace{0.5cm}\Rightarrow\hspace{0.5cm}
E^2 - 198.0 E + 6602.1 = 0
\tag{A.18}
\end{equation}
and solve for the roots of the quadratic equation; the two solutions (in units of eV) are $E^{(1)} = 42.441$ and $E^{(2)} = 155.56$. These are the eigenvalues of the Hamiltonian matrix and correspond to the allowed energies for an electron that experiences the potential given in Eq. (A.1).

Lastly, we use these eigenvalues to determine the eigenvectors of ${\bf H}$. First, we insert $E^{(1)} = 42.441$ into the matrix form of the Schrödinger equation as in Eq. (A.9) with the Hamiltonian given in Eq. (A.16)

\begin{equation}
\begin{pmatrix}
42.6 & -4.24 \\
-4.24 & 155.4 
\end{pmatrix}
\begin{pmatrix}
c_1^{(1)} \\ c_2^{(1)}
\end{pmatrix} 
= 42.441
\begin{pmatrix}
c_1^{(1)} \\ c_2^{(1)}
\end{pmatrix}
\hspace{0.5cm}\Rightarrow\hspace{0.5cm}
\begin{pmatrix}
42.6 c_1^{(1)}  -  4.24 c_2^{(1)} \\
-4.24 c_1^{(1)} + 155.4 c_2^{(1)} 
\end{pmatrix}=
\begin{pmatrix}
42.441 c_1^{(1)} \\ 42.441 c_2^{(1)}
\end{pmatrix}
\tag{A.19}
\end{equation}
The matrix equation (2.19) is equivalent two coupled equations, the first one being $42.6 c^{(1)}_1  -  4.24 c_2^{(1)} = 42.441 c_1^{(1)}$. Solving for $c_1^{(1)}$ in terms of $c_2^{(1)}$, we get 
\begin{equation}
c_1^{(1)} = 26.667 c_2^{(1)}
\tag{A.20}
\end{equation}
Since the basis functions $\phi_1(x)$ and $\phi_2(x)$ are orthonormal (see Eq. (A.3)), one can show that for a normalized wave function $\int_0^L \vert \psi(x)  \vert^2 {\rm d}x = 1$, the coefficients $c_1^{(1)}$ and $c_2^{(1)}$ are related by
\begin{equation}
\left(c_1^{(1)}\right)^2 + \left(c_2^{(1)}\right)^2 = 1
\tag{A.21}
\end{equation}
Inserting Eq. (2.20) into Eq. (2.21) gives
\begin{equation}
\left( 3.65833 c_2^{(1)} \right)^2 + \left(c_2^{(1)}\right)^2 = 1
\hspace{0.5cm}\Rightarrow\hspace{0.5cm}
\left( c_2^{(1)}\right) ^2 = 1.404\times10^{-3} 
\hspace{0.5cm}\Rightarrow\hspace{0.5cm}
c_2^{(1)} = \pm 0.03747 
\tag{A.22}
\end{equation}
Because it is the relative sign of $c_1^{(1)}$ and $c_2^{(1)}$ that matters, we can pick either solution in Eq. (A.22). Using the positive solution, we get $c_1^{(1)} = 0.99860$. Thus, the first eigenpair is given by
\begin{equation}
E^{(1)}=42.441 \hspace{1cm}\&\hspace{1cm} {\bf c}^{(1)} = 
\begin{pmatrix}
0.99860 \\ 0.03747
\end{pmatrix}
\hspace{1cm}\Rightarrow\hspace{1cm} \psi^{(1)}(x) = 0.99860\cdot \phi_1(x) + 0.03747\cdot \phi_2(x)
\tag{A.23}
\end{equation}
Using similar logic, we get the second eigenpair
\begin{equation}
E^{(2)}=155.56 \hspace{1cm}\&\hspace{1cm} {\bf c}^{(2)} = 
\begin{pmatrix}
- 0.03747 \\  0.99860
\end{pmatrix}
\hspace{1cm}\Rightarrow\hspace{1cm} \psi^{(2)}(x) = - 0.03747\cdot \phi_1(x) + 0.99860\cdot \phi_2(x)
\tag{A.24}
\end{equation}

Inspection of the Hamiltonian matrix in Eq. (A.16) reveals that it is symmetric since $H_{1,2} = H_{2,1}$. In general, Hamiltonian matrices with real elements are always symmetric and $H_{m,n} = H_{n,m}$. As it turns out, finding the eigenvectors of a matrix is equivalent to diagonalizing a matrix. Diagonalization is defined as finding the similarity transformation matrix ${\bf P}$ that converts the matrix $\bf A$ to a diagonal matrix ${\bf D} = {\bf P}^T {\bf A}{\bf P}$ where the columns of $\bf P$ correspond to the eigenvectors of the matrix $\bf A$ and $^T$ denotes the transpose (obtained by interchanging colum and row indeces) of a matrix. To demonstrate this, using the two eigenvectors found in Eqs. (A.23) and (A.24), we have
\begin{equation}
{\bf P} =  
\begin{pmatrix} {\bf c}^{(1)} & {\bf c}^{(2)} \end{pmatrix} = 
\begin{pmatrix}
0.99860 & - 0.03747 \\
0.03747 &   0.99860
\end{pmatrix}
\hspace{1cm}\Rightarrow\hspace{1cm} 
{\bf P}^T =
\begin{pmatrix}
  0.99860 &   0.03747 \\
- 0.03747 &   0.99860
\end{pmatrix}
\end{equation}

For the Hamiltonian in Eq. (A.16) we see that
\begin{eqnarray}
{\bf D} = {\bf P}^T{\bf H}{\bf P} &=&
\begin{pmatrix}
  0.99860 &   0.03747 \\
- 0.03747 &   0.99860
\end{pmatrix}
\begin{pmatrix}
  42.6 & -4.24 \\
-4.24 &  155.4 
\end{pmatrix}
\begin{pmatrix}
0.99860 & - 0.03747 \\
0.03747 &   0.99860
\end{pmatrix}
\\
&=&
\begin{pmatrix}
  0.99860 &   0.03747 \\
- 0.03747 &   0.99860
\end{pmatrix}
\begin{pmatrix}
42.382 & -5.830 \\
1.589 &  155.34
\end{pmatrix} 
\\
&=&
\begin{pmatrix}
42.382 &  -0.001 \\
 -0.001 & 155.34
\end{pmatrix} 
\end{eqnarray}

As demonstrated in the cell below (go ahead, and run the code yourself), the fact that the off-diagonal elements of the above matrix are not exactly zero and that the diagonal elements are slightly different from the eigenvalues of ${\bf H}$ *is the result of rounding*. Nonetheless, this example demonstrates that finding the matrix $\bf P$ that brings $\bf H$ to diagonal form is equivalent to finding the eigenvalues and eigenvectors or $\bf H$. 

In [1]:
import numpy as np
import scipy as sp

def ratio_value(H_11,H_12,E):
    return H_12/(E-H_11)

def c2_value(ratio):
    return np.sqrt(1.0/(1.0+ratio**2))

def c1_value(ratio,c2):
    return ratio * c2

def root_value(a,b,c,root):
    print(a*root**2+b*root+c)

# H matrix elements
H_11 = 42.6
H_12 = -4.24
H_22 = 155.4

# coefficients in quadratic equation
a    = 1.0
b    = -(H_11+H_22)
c    = +H_11*H_22-(H_12**2)

# roots of quadratic equation
E1   = ( - b - np.sqrt(b**2 - 4.0*a*c) ) / ( 2.0 * a )
E2   = ( - b + np.sqrt(b**2 - 4.0*a*c) ) / ( 2.0 * a )

print("")
# solve for coefficients for eigenvalue E1
ratio_1 = ratio_value(H_11,H_12,E1)
c_21    = c2_value(ratio_1)
c_11    = c1_value(ratio_1,c_21)
print(F'E1={E1:12.8f} c_1={c_11:12.8f} c_2={c_21:12.8f}')

# solve for coefficients for eigenvalue E2
ratio_2 = ratio_value(H_11,H_12,E2)
c_22    = c2_value(ratio_2)
c_12    = c1_value(ratio_2,c_22)
print(F'E2={E2:12.8f} c_1={c_12:12.8f} c_2={c_22:12.8f}')

# define Hamiltonian matrix
H = [[H_11,H_12],[H_12,H_22]]
# construct similarity transformation matrix P 
P = [ [c_11 , c_12] , [c_21 , c_22] ]
print("")
print("P = ")
print(P)

#bring Hamiltonian to diagonal form
print("")
print("P^T H P =")
print(np.matmul(np.transpose(P),np.matmul(H,P)))

# much much easier to do with Python's diagonalization routine
eig,vec = sp.linalg.eig(H)

print("")
print("eigenvalues after diagonalization")
print(np.real(eig))

print("")
print("eigenvectors after diagonalization (note that overall phases might differ from the ones above)")
print(vec)


E1= 42.44084866 c_1=  0.99929628 c_2=  0.03750928
E2=155.55915134 c_1= -0.03750928 c_2=  0.99929628

P = 
[[0.9992962794157921, -0.03750927812895497], [0.03750927812895414, 0.9992962794157921]]

P^T H P =
[[ 4.24408487e+01 -1.29064543e-13]
 [-1.29269324e-13  1.55559151e+02]]

eigenvalues after diagonalization
[ 42.44084866 155.55915134]

eigenvectors after diagonalization (note that overall phases might differ from the ones above)
[[-0.99929628  0.03750928]
 [-0.03750928 -0.99929628]]
